### PR TITLE
Don't use ACL names ever!

### PR DESCRIPTION
**Pull request by tssurya**
Initially when the commit to cleanup ACLs was done in https://github.com/ovn-org/ovn-kubernetes/pull/3038 I had assumed that ACL names weren't truncated to fit the 63 char length and thought they were like other fields. Based on that assumption there were two places in code where I used a predicate match based on acl names which was plain stupidity. This PR fixes that:

1) When the arp && arp || nd bug was being cleaned up, I was trying to be extra careful in matching on old exp: "arp" and the acl name. I deleted the second part of that logic since there can be cases where the entire suffix is missing like when the namespace is 63 chars. So let us stick to matching on acl.Match alone

2) When the deny default ACL duplicates per namespace were being updated - again I was matching on ACL names to exclude the arp policies. I replaced this with using acl.Match which is more reliable and accurate.

In https://github.com/ovn-org/ovn-kubernetes/pull/3181 we realised we had to trim the acls and we fixed that, but didn't fix the above two issues. So let's fix that up and also I added a unit test case with a long named namespace.

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

/assign @jcaamano

### CODE REVIEW

File: go-controller/pkg/ovn/policy.go [hunk 1]
```diff
@@ -195,7 +195,7 @@ func (oc *Controller) updateStaleDefaultDenyACLNames(npType knet.PolicyType, gre
 	p := func(item *nbdb.ACL) bool {
 		return item.ExternalIDs[defaultDenyPolicyTypeACLExtIdKey] == string(npType) && // default-deny-policy-type:Egress or default-deny-policy-type:Ingress
 			strings.Contains(item.Match, gressSuffix) && // Match:inport ==	@ablah80448_egressDefaultDeny or Match:inport == @ablah80448_ingressDefaultDeny
-			!strings.Contains(*item.Name, arpAllowPolicySuffix) && // != name: namespace_ARPallowPolicy
+			!strings.Contains(item.Match, "arp") && // != match: (arp || nd)
```
Comment (jcaamano) [inline go-controller/pkg/ovn/policy.go:198]: Why not define a constant for the full match that is used both there and where the ACL is created. That way if it is changed in one place in the future, we won't miss the change here and will be reminded to check if this logic is still valid.

Comment (jcaamano) [inline go-controller/pkg/ovn/policy.go:198]: To discard already migrated ACLs, why not filter out names that end with a substring of `gressSuffix` after the `_`. Otherwise we would always be migrating the deny ACLs with long namespace names.

Comment (jcaamano) [inline go-controller/pkg/ovn/policy.go:198]: Something we forgot about the first time around, set the ACL UUID to make the operation more performant:
`newACL.UUID := aclList[0].UUID`

File: go-controller/pkg/ovn/policy_test.go [hunk 2]
```diff
@@ -1968,6 +1968,251 @@ var _ = ginkgo.Describe("OVN NetworkPolicy Operations", func() {
...
+
+				egressOptions := map[string]string{
+					// older versions of ACLs don't have, should be added by syncNetworkPolicies on startup
+					//	"apply-after-lb": "true",
+				}
+				longLeftOverNameSpaceName := "abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz"
+				pgHash := hashedPortGroup(longLeftOverNameSpaceName)
+				// ACL1: leftover arp allow ACL egress with old match (arp)
+				leftOverACL1FromUpgrade := libovsdbops.BuildACL(
+					longLeftOverNameSpaceName+"_ARPallowPo",
```
Comment (jcaamano) [inline go-controller/pkg/ovn/policy_test.go:2042]: `BuildACL` will already trim the name so you can use the constants to build the full name here.

File: go-controller/pkg/ovn/policy.go (hunk 1 above)
Comment (tssurya) [reply go-controller/pkg/ovn/policy.go:198]: great point Jaime! let me use constants..

Comment (tssurya) [reply go-controller/pkg/ovn/policy.go:198]: yeah I know we will be re-doing some of the work always on restart, but I felt that was a small price to pay since I thought it would be a corner case. TBH I am very hesitant to add back any filtering based on names or suffixes that go into the names. Example for the corner case 63 we will always redo migration for those acls.

Let me think a bit more if this is worth the complexity or not. But I have been thinking of this since yesterday... Maybe the match after `_` is ok since I checked that in K8s one cannot have a `_` in a namespace name. And along with the other three matches... maybe I am ok to consider the names substring match after the `_`. I wonder if just checking for `_` is good enough? Because since we rule out all arp allow acls and the first two match conditions will ensure we don't pick non-default acls.. that might work. @jcaamano : shall we just do:
1) a `_` match?
2) or does it really add value when we match on the `_e`, `_eg`, `_egre` and so on?
3) or we decide this is corner case maybe 10 namespaces in the cluster will always be migrated on restart?

Comment (tssurya) [reply go-controller/pkg/ovn/policy.go:198]: oh! ack, so if I understand this correctly when I set the UUID while giving it for update, since UUID is a unique identifier, the indexing on it will help in having a more efficient fetchACL algorithm?

Comment (tssurya) [reply go-controller/pkg/ovn/policy.go]: update: let me push an update for this and we can talk post that.

Comment (jcaamano) [reply go-controller/pkg/ovn/policy.go]: The last `_` should be the one that we added provided there is no other `_` in the suffix. Also any other suffix we use for these ACLs should not share any common sub-string with this suffix. Both conditions hold true right now and we can probably document where we are defining them the matter that we rely on these facts.

If the namespace name ends with our suffix sub-string and is at the maximum length then it would be a noop in any case.

We can also test if the new name is equal to the old name and omit the update in that case.

I think that the most easy thing to do is to check that our suffix is a prefix of the ACL name suffix.


If we suspect a corner case we can test for it.

Comment (tssurya) [reply go-controller/pkg/ovn/policy.go]: sounds good, I have fixed this. Let me add some tests for the prefix in suffix logic that is standalone.

Comment (tssurya) [reply go-controller/pkg/ovn/policy.go]: done!

**Comment by tssurya** [comment]
hmm need to rebase... something merged here

**Comment by tssurya** [comment]
FYI: Anurag did an initial testing around this and it seems to pass. However we need to provide the fix more officially.

File: go-controller/pkg/ovn/policy.go [hunk 3]
```diff
@@ -185,24 +191,40 @@ func hashedPortGroup(s string) string {
...
+		if len(aclNameSuffix) == 1 {
+			// doesn't have suffix; no update required; append the actual suffix since this ACL can be skipped
+			aclNameSuffix = append(aclNameSuffix, gressSuffix)
+		}
 		return item.ExternalIDs[defaultDenyPolicyTypeACLExtIdKey] == string(npType) && // default-deny-policy-type:Egress or default-deny-policy-type:Ingress
 			strings.Contains(item.Match, gressSuffix) && // Match:inport ==	@ablah80448_egressDefaultDeny or Match:inport == @ablah80448_ingressDefaultDeny
-			!strings.Contains(*item.Name, arpAllowPolicySuffix) && // != name: namespace_ARPallowPolicy
-			!strings.Contains(*item.Name, gressSuffix) // filter out already converted ACLs (we still continue to pick the cases where names were >45 chars)
+			!strings.Contains(item.Match, staleArpAllowPolicyMatch) && // != match: (arp || nd) OR arp
+			!prefixes.Has(aclNameSuffix[1]) // filter out already converted ACLs or ones that are a no-op
```
Comment (jcaamano) [inline go-controller/pkg/ovn/policy.go:218]: we can just use here `strings.HasPrefix(gressSuffix, aclNameSuffix)`, we don't need `generatePrefix` I think

File: go-controller/pkg/ovn/policy.go [hunk 4]
```diff
@@ -185,24 +191,40 @@ func hashedPortGroup(s string) string {
...
+		aclNameSuffix := strings.Split(*item.Name, "_")
+		if len(aclNameSuffix) == 1 {
+			// doesn't have suffix; no update required; append the actual suffix since this ACL can be skipped
+			aclNameSuffix = append(aclNameSuffix, gressSuffix)
+		}
 		return item.ExternalIDs[defaultDenyPolicyTypeACLExtIdKey] == string(npType) && // default-deny-policy-type:Egress or default-deny-policy-type:Ingress
 			strings.Contains(item.Match, gressSuffix) && // Match:inport ==	@ablah80448_egressDefaultDeny or Match:inport == @ablah80448_ingressDefaultDeny
-			!strings.Contains(*item.Name, arpAllowPolicySuffix) && // != name: namespace_ARPallowPolicy
-			!strings.Contains(*item.Name, gressSuffix) // filter out already converted ACLs (we still continue to pick the cases where names were >45 chars)
+			!strings.Contains(item.Match, staleArpAllowPolicyMatch) && // != match: (arp || nd) OR arp
```
Comment (jcaamano) [inline go-controller/pkg/ovn/policy.go:217]: The `staleArpAllowPolicyMatch` ACLs are deleted later on the sync, so it seems that we would not have to account for them here if we moved `updateStaleDefaultDenyACLNames` after that section. 
In general, in the sync section, we should delete ACLs first so that they are not considered in later predicates.
But this is how it was before, so up to you.

File: go-controller/pkg/ovn/policy.go [hunk 5]
```diff
@@ -321,8 +344,7 @@ func (oc *Controller) syncNetworkPolicies(networkPolicies []interface{}) error {
 	// remove stale egress and ingress allow arp ACLs that were leftover as a result
 	// of ACL migration for "ARPallowPolicy" when the match changed from "arp" to "(arp || nd)"
 	p = func(item *nbdb.ACL) bool {
-		return strings.Contains(item.Match, " && arp") &&
-			strings.Contains(*item.Name, arpAllowPolicySuffix)
+		return strings.Contains(item.Match, " && "+staleArpAllowPolicyMatch)
```
Comment (jcaamano) [inline go-controller/pkg/ovn/policy.go:347]: Since we are not matching by suffix, would we fill more comfortable if we additionally matched by the external id? Assuming the external id should be there....

Comment (tssurya) [reply go-controller/pkg/ovn/policy.go:347]: Adding externalID here doesn't help us much since the same externalID is there for all default ingress ACLs. I thought since the one thing we are certain about is the `arp` match being wrong, let's go only by that. Do we have any value in adding the externalID here?

File: go-controller/pkg/ovn/policy.go (hunk 3 above)
Comment (tssurya) [reply go-controller/pkg/ovn/policy.go:218]: the `HasPrefix` checks if the latter string provided is a prefix of the former, so the whole string would be matched as a prefix which in our case is the full `egressDefaultDeny`, we won't be accouting for the trimmed cases like `egress` or `egr` etc.. at least from reading the golang utility page for `HasPrefix` that is my understanding of how it works.

File: go-controller/pkg/ovn/policy.go (hunk 4 above)
Comment (tssurya) [reply go-controller/pkg/ovn/policy.go:217]: ack! this is something I can change. I am adding some more corner case tests into the unit test, let me push this and that together.

File: go-controller/pkg/ovn/policy.go (hunk 3 above)
Comment (jcaamano) [reply go-controller/pkg/ovn/policy.go:218]: How is that?
`strings.HasPrefix("egressDefaultDeny", "egress")` would be true?

> the HasPrefix checks if the latter string provided is a prefix of the former
correct

>  so the whole string would be matched as a prefix which in our case is the full egressDefaultDeny, we won't be accouting for the trimmed cases like egress or egr etc.. at least from reading the golang utility page for HasPrefix that is my understanding of how it works

no, in `strings.HasPrefix(gressSuffix, aclNameSuffix)`, `aclNameSuffix` is the (probably trimmed) suffix the ACL name has and `gressSuffix` is the full suffix constant so

the full case: `strings.HasPrefix("egressDefaultDeny", "egressDefaultDeny")` would be true as "egressDefaultDeny" is a prefix of "egressDefaultDeny"
the trimmed case: `strings.HasPrefix("egressDefaultDeny", "egress")` would be true as "egress" is a prefix of "egressDefaultDeny"

what am I missing?

File: go-controller/pkg/ovn/policy.go (hunk 5 above)
Comment (jcaamano) [reply go-controller/pkg/ovn/policy.go:347]: the same externalID is there for all default deny ACLs but not all ACLs. So we would be matching "arp" only within default deny ACLs and not all ACLs. It has the same use as the one we are giving it in this other predicate we use in line 215
```
		return item.ExternalIDs[defaultDenyPolicyTypeACLExtIdKey] == string(npType) && // default-deny-policy-type:Egress or default-deny-policy-type:Ingress
			strings.Contains(item.Match, gressSuffix) && // Match:inport ==	@ablah80448_egressDefaultDeny or Match:inport == @ablah80448_ingressDefaultDeny
			!strings.Contains(*item.Name, arpAllowPolicySuffix) && // != name: namespace_ARPallowPolicy
			!strings.Contains(*item.Name, gressSuffix) // filter out already converted ACLs (we still continue to pick the cases where names were >45 chars)
			!strings.Contains(item.Match, staleArpAllowPolicyMatch) && // != match: (arp || nd) OR arp
```

File: go-controller/pkg/ovn/policy.go (hunk 3 above)
Comment (tssurya) [reply go-controller/pkg/ovn/policy.go:218]: oh got it! sorry went off tangent

File: go-controller/pkg/ovn/policy.go (hunk 5 above)
Comment (tssurya) [reply go-controller/pkg/ovn/policy.go:347]: yeah better safe than sorry, I will use externalIDs

**Comment by tssurya** [comment]
/retest-failed
I am unsure why junit is reporting failures when the unit tests pass..

**Review by jcaamano — APPROVED** [review]
